browser/tor/tor_control.cc

```diff
@@ -0,0 +1,1512 @@
+            std::string key, value;
+            if (!ParseKV(reply, &key, &value)) {
+              VLOG(1) << "tor: invalid async event";
+              Error();
+              return false;
+            }
+            if (async_->extra.count(key)) {
+              VLOG(1) << "tor: duplicate key in async event";
+              Error();
+              return false;
+            }
+            async_->extra[key] = value;
+
+            // If we're still subscribed, notify the delegate of the
+            // parsed reply.
+            if (async_events_.count(async_->event)) {
+              NotifyTorEvent(
+                    async_->event, async_->initial, async_->extra);
+            }
+          }
+          async_.reset();
+          return true;
+        }
+      }
+    }
+  } else {
+    // Synchronous reply.  Return it to the next command callback in
+    // the queue.
+    switch (pos) {
+      case '-':
+        NotifyTorRawMid(status, reply);
+        if (!cmdq_.empty()) {
+          PerLineCallback& perline = cmdq_.front().first;
+          perline.Run(status, reply);
+        }
+        return true;
+      case '+':
+        VLOG(2) << "tor: NYI: control data reply";
+        // XXX Just ignore it for now.
+        return true;
+      case ' ':
+        NotifyTorRawEnd(status, reply);
+        if (!cmdq_.empty()) {
+          CmdCallback& callback = cmdq_.front().second;
+          bool error = false;
+          std::move(callback).Run(error, status, reply);
+          cmdq_.pop();
+        }
+        return true;
+    }
+  }
+
+  // Not reached if the line is well-formed.
+  VLOG(1) << "tor: malformed control line: "
+             << escapify(line.data(), line.size());
+  Error();
+  return false;
+}
+
+TorControl::Async::Async() = default;
+TorControl::Async::~Async() = default;
+
+// Error()
+//
+//      Clear read and write state and disconnect.
+//
+void TorControl::Error() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(io_sequence_checker_);
+
+  VLOG(1) << "tor: closing control on " << (running_ ? "request" : "error");
+
+  NotifyTorClosed();
+
+  // Invoke all callbacks with errors and clear read state.
+  while (!cmdq_.empty()) {
+    CmdCallback& callback = cmdq_.front().second;
+    bool error = true;
+    std::move(callback).Run(error, "", "");
+    cmdq_.pop();
+  }
+  reading_ = false;
+  readiobuf_.reset();
+  read_start_ = -1;
+  read_cr_ = false;
+
+  // Clear write state.
+  writeq_ = {};
+  writing_ = false;
+  writeiobuf_.reset();
+
+  // Clear the socket.
+  socket_.reset();
+
+  // If we're still running, try watching again to start over.
+  //
+  // XXX Rate limit in case of flapping?
+  if (running_) {
+    watch_task_runner_->PostTask(
+        FROM_HERE,
+        base::BindOnce(&TorControl::Poll, base::Unretained(this)));
+  }
+}
+
+void TorControl::NotifyTorControlReady() {
+  owner_task_runner_->PostTask(
+      FROM_HERE,
+      base::BindOnce(
+          [](TorControl::Delegate* delegate) {
```

Comment:
why are we using lambdas here? owner_task_runner_ is the UI thread, correct? I think these can potentially trigger a crash on shutdown

Reply:
also since `OnTorControlReady` explicitly expects to be called on the UI thread I think we should explicitly post to the UI thread instead of using `owner_task_runner_`. Or alternatively `OnTorControlReady` should expect to be called on `owner_task_runner_` instead of explicitly checking the UI thread, but I'm not seeing any benefit to the latter

Reply:
fixed in d3723d590c318b18927c45bb004c2e580d48054c